src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2596,7 +2596,7 @@ object ICustomAstVisitor2.VisitPipelineChain(PipelineChainAst pipelineChainAst)
             var types = new List<PSTypeName>();
             types.AddRange(InferTypes(pipelineChainAst.LhsPipelineChain));
             types.AddRange(InferTypes(pipelineChainAst.RhsPipeline));
-            return GetArrayType(types);
+            return types.Distinct();
```

Comment:
Hmm, I see GetArrayType() trying process IEnumerable. It looks reasonable. So maybe improve GetArrayType() method itself?
Have we tests for scenarios with IEnumerable?

Reply:
GetArrayType is simply the wrong choice for this purpose so there is nothing to improve. It's goal is to find the most specific array type so if you had something like: `$Guid = Get-SomeGuid || New-Guid` it would infer the Guid as `System.Guid[]` or if you had something like: `$DateOrTimespan = Get-SomeDate || New-TimeSpan` it would infer it as `System.Object[]`.

In every other type inference scenario I can think of, we output multiple types if there are multiple possible outcomes. An example is if we infer `Get-ChildItem C:\` it's inferred as both FileInfo and DirectoryInfo and tab completion shows members from both types.

Reply:
I see you point. I think it works better than before.